src/sbd-cluster.c

```diff
@@ -67,21 +70,137 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
+static cmap_track_handle_t track_handle = 0;
+static int cpg_membership_entries = -1;
+static GSource *cmap_source = NULL;
+
+void
+sbd_cpg_membership_health_update()
+{
+    if(cpg_membership_entries > 0) {
+        if (!two_node || !ever_seen_both || cpg_membership_entries > 1) {
+            set_servant_health(pcmk_health_online, LOG_INFO,
+                           "Connected to %s (%u members)",
+                           name_for_cluster_type(get_cluster_type()),
+                           cpg_membership_entries
+                          );
+        } else {
+            /* should ask votequorum but like this races between
+             * votequorum & membership callbacks are no issue
+             * and no bindings to votequorum api are needed
+             */
+            set_servant_health(pcmk_health_noquorum, LOG_WARNING,
+                           "Connected to %s but requires both nodes present",
+                           name_for_cluster_type(get_cluster_type())
+                          );
+        }
+
+        if (cpg_membership_entries > 1) {
+            ever_seen_both = true;
+        }
+    } else {
+        set_servant_health(pcmk_health_unclean, LOG_WARNING,
+                           "Empty %s membership", name_for_cluster_type(get_cluster_type()));
+    }
+}
+
 void
 sbd_cpg_membership_dispatch(cpg_handle_t handle,
                     const struct cpg_name *groupName,
                     const struct cpg_address *member_list, size_t member_list_entries,
                     const struct cpg_address *left_list, size_t left_list_entries,
                     const struct cpg_address *joined_list, size_t joined_list_entries)
 {
-    if(member_list_entries > 0) {
-        set_servant_health(pcmk_health_online, LOG_INFO,
-                           "Connected to %s", name_for_cluster_type(get_cluster_type()));
+    cpg_membership_entries = member_list_entries;
+    sbd_cpg_membership_health_update();
+    notify_parent();
+}
+
+static void sbd_cmap_notify_fn(
+    cmap_handle_t cmap_handle,
+    cmap_track_handle_t cmap_track_handle,
+    int32_t event,
+    const char *key_name,
+    struct cmap_notify_value new_val,
+    struct cmap_notify_value old_val,
+    void *user_data)
+{
+    if ((event == CMAP_TRACK_MODIFY) &&
+       (new_val.type == CMAP_VALUETYPE_UINT8)) {
+        two_node = *((uint8_t *) new_val.data);
+        sbd_cpg_membership_health_update();
+        notify_parent();
+    }
+}
+
+static gboolean
+cmap_dispatch_callback (gpointer user_data)
+{
+    cmap_dispatch(cmap_handle, CS_DISPATCH_ALL);
+    return TRUE;
+}
+
+static gboolean
+sbd_get_two_node(void)
+{
+    uint8_t two_node_u8 = 0;
+    int cmap_fd;
+
+    if (!track_handle) {
+        if (cmap_initialize(&cmap_handle) != CS_OK) {
+            cl_log(LOG_WARNING, "Cannot initialize CMAP service\n");
+            goto out;
+        }
+
+        if (cmap_track_add(cmap_handle, "quorum.two_node", CMAP_TRACK_MODIFY,
```

Comment:
Tracking just modify operation is not enough. You have to track also add and delete operations (CMAP_TRACK_ADD and CMAP_TRACK_DELETE)